### PR TITLE
Memory-handling for feathering large cubes

### PR DESCRIPTION
**Pull request by e-koch**
Continuation of #25.

* [x] Add memmap feature
* [x] Dask-enabled cube feathering
* [x] Add docs page on using memmap and for enabling dask use
* [x] Also add general cube feathering docs.
* [x] Handling for varying res cube (but NOT with dask yet)

### CODE REVIEW

**Comment by e-koch** [comment]
@keflavich -- This is ready for a first review. VaryingRes cubes aren't handle yet, and there's a couple places where the memory handling might break (needing to reproject the SD cube). But the critical tests pass and it's in good enough shape to start stress-testing locally.

File: docs/feathering_cubes.rst [hunk 1]
```diff
@@ -3,14 +3,60 @@
...
+    >>> feathered_cube = feather_simple_cube(highres_cube, lowres_cube)  # doctest: +SKIP
+
+This will automatically enable using dask mode in `~uvcombine.feather_simple_cube`.
+
+A required pre-processing step is to reproject `lowres_cube` to the same pixel
+grid as `highres_cube` and match the brightness units. As this can be computationally
+expensive, the spectral-cube dask integration exposes an option to save temporary
+intermediate products as zarr files. This is especially useful when rechunking the
+data to optimize different computations (e.g., spectral versus spatial regridding.)
+To enable this mode, `use_save_to_tmp_dir` can be enabled in `~uvcombine.feather_simple_cube`.
```
Comment (keflavich) [inline docs/feathering_cubes.rst:46]: Link to https://spectral-cube.readthedocs.io/en/latest/dask.html#saving-intermediate-results-to-disk

Comment (keflavich) [reply docs/feathering_cubes.rst:46]:
```suggestion
To enable this mode, [`use_save_to_tmp_dir`](https://spectral-cube.readthedocs.io/en/latest/dask.html#saving-intermediate-results-to-disk) can be enabled in `~uvcombine.feather_simple_cube`.
```

or... how do you do it with intersphinx?  I can't remember.  Argh.

File: docs/feathering_cubes.rst [hunk 2]
```diff
@@ -3,14 +3,60 @@
...
+This will additionally trigger saving the feathered cube to a temporary zarr file and
+avoids delaying the feathering computation::
```
Comment (keflavich) [inline docs/feathering_cubes.rst:48]: "avoids delaying" sounds like delaying is something you want to do, but I think it's not.  Maybe rephrase to, e.g.,
"and performs the feathering computation immediately instead of delaying it (delayed computation is the default when using dask arrays)."  

But... that adds complexity here.  Maybe it's best not to bring this up unless we add a separate section about delayed computation

File: uvcombine/tests/test_feather.py [hunk 3]
```diff
@@ -90,9 +114,13 @@ def test_feather_simple_mismatchunit(plaw_test_data):
...
 
-def test_fourier_combine_cubes(plaw_test_cube_sc):
+def test_fourier_combine_cubes(cube_data):
 
-    orig_cube, sd_cube, interf_cube = plaw_test_cube_sc
+    orig_fname, sd_fname, interf_fname = cube_data
+
+    orig_cube, orig_data = cube_and_raw(orig_fname, use_dask=False)
+    sd_cube, sd_data = cube_and_raw(sd_fname, use_dask=False)
+    interf_cube, interf_data = cube_and_raw(interf_fname, use_dask=False)
```
Comment (keflavich) [inline uvcombine/tests/test_feather.py:123]: any reason `use_dask` isn't looped over here?

File: uvcombine/uvcombine.py [hunk 4]
```diff
@@ -973,9 +973,63 @@ def spectral_smooth_and_downsample(cube, kernelfwhm):
...
+                            highresscalefactor=1.0,
+                            lowresscalefactor=1.0,
+                            weights=1.0,
+                            replace_hires=False,
+                            lowpassfilterSD=False,
+                            deconvSD=False):
+
+        lowresfwhm = cube_lo.beam.major
+
+        pixscale = wcs.utils.proj_plane_pixel_scales(cube_hi.wcs.celestial)[0]
```
Comment (keflavich) [inline uvcombine/uvcombine.py:997]: this pattern annoys me, so:
https://github.com/astropy/astropy/pull/13268

File: uvcombine/uvcombine.py [hunk 5]
```diff
@@ -1002,39 +1075,110 @@ def feather_simple_cube(cube_hi, cube_lo,
...
+        # Check kwargs for feather_simple kwarg to allow matching units
+        match_units = kwargs.pop('match_units', True)
+
+        # Check for units consistency
+        if match_units:
+            # After this step, the units of cube_hi are some sort of surface brightness
+            # unit equivalent to that specified in the high-resolution header's units
+
+            # NOTE: is this memory-friendly??
+            cube_lo = cube_lo.to(cube_hi.unit)
```
Comment (keflavich) [inline uvcombine/uvcombine.py:1138]: probably not but I wish it was; I do this all the time.

Comment (keflavich) [reply uvcombine/uvcombine.py:1138]: this should be `cube_lo_reproj`
```suggestion
            # NOTE: is this memory-friendly??
            cube_lo_reproj = cube_lo_reproj.to(cube_hi.unit)
```

File: uvcombine/uvcombine.py [hunk 6]
```diff
@@ -1002,39 +1075,110 @@ def feather_simple_cube(cube_hi, cube_lo,
...
+
+            # When in a per-beam unit, we need to scale the low res to the
+            # Jy / beam for the HIRES beam.
+            jybm_unit = u.Jy / u.beam
+            if cube_hi.unit.is_equivalent(jybm_unit):
+                cube_lo *= (cube_hi.beam.sr / cube_lo.beam.sr).decompose().value
```
Comment (keflavich) [inline uvcombine/uvcombine.py:1144]: ditto
```suggestion
            if cube_hi.unit.is_equivalent(jybm_unit):
                cube_lo_reproj *= (cube_hi.beam.sr / cube_lo_reproj.beam.sr).decompose().value
```
this is important also b/c `*=` could look like it was going to modify user input if this weren't already a copy

File: uvcombine/uvcombine.py [hunk 7]
```diff
@@ -1002,39 +1075,110 @@ def feather_simple_cube(cube_hi, cube_lo,
...
+
+        # Add check that the units are compatible
+        equiv_units = cube_lo.unit.is_equivalent(cube_hi.unit)
```
Comment (keflavich) [inline uvcombine/uvcombine.py:1147]:
```suggestion
        equiv_units = cube_lo_reproj.unit.is_equivalent(cube_hi.unit)
```

File: uvcombine/uvcombine.py [hunk 8]
```diff
@@ -1002,39 +1075,110 @@ def feather_simple_cube(cube_hi, cube_lo,
...
+        if not equiv_units:
+            raise ValueError("Brightness units are not equivalent: "
+                            f"hires: {cube_hi.unit}; lowres: {cube_lo.unit}")
+
+        feathcube = _dask_feather_cubes(cube_hi, cube_lo,
```
Comment (keflavich) [inline uvcombine/uvcombine.py:1152]:
```suggestion
        feathcube = _dask_feather_cubes(cube_hi, cube_lo_reproj,
```

File: uvcombine/uvcombine.py [hunk 9]
```diff
@@ -1002,39 +1075,110 @@ def feather_simple_cube(cube_hi, cube_lo,
...
+        else:
+            feath_array = np.empty(cube_hi.shape)
 
-        feath_array[ii] = feather_simple(hslc, lslc, **kwargs)
+        pb = ProgressBar(cube_hi.shape[0])
+        for ii in range(cube_hi.shape[0]):
 
-        pb.update()
+            hslc = cube_hi[ii]
+            lslc = cube_lo[ii]
```
Comment (keflavich) [inline uvcombine/uvcombine.py:1169]: I think here too?  I may have lost which function we're in though
```suggestion
            lslc = cube_lo_reproj[ii]
```

**Review by keflavich — CHANGES_REQUESTED** [review]

File: uvcombine/tests/test_feather.py (hunk 3 above)
Comment (e-koch) [reply uvcombine/tests/test_feather.py:123]: Uses the old cube feather function that doesn't have dask handling.

File: uvcombine/uvcombine.py (hunk 9 above)
Comment (e-koch) [reply uvcombine/uvcombine.py:1169]: Only needed fixing for the dask version. Otherwise `feather_simple` handles this per plane.

File: uvcombine/uvcombine.py (hunk 4 above)
Comment (e-koch) [reply uvcombine/uvcombine.py:997]: I'll add an issue here for this as we use this EVERYWHERE

Comment (e-koch) [reply uvcombine/uvcombine.py:997]: #35

File: docs/feathering_cubes.rst (hunk 2 above)
Comment (e-koch) [reply docs/feathering_cubes.rst]: Removed this part to sidestep the longer explanation of delayed computation.

**Comment by e-koch** [comment]
@keflavich -- Added some more checks and such. I think this is feature complete now and is at least running on ~few GB cubes for feathering, w/ and w/o dask.

Dask VRSC isn't handled yet. I think that should be revisited later to get these adds wrapped up.

**Review by keflavich — APPROVED** [review]
